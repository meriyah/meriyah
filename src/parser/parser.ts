import { convertTokenType } from '../lexer';
import { Token } from '../token';
import type * as ESTree from '../estree';
import {
  Flags,
  type OnComment,
  type OnInsertedSemicolon,
  type OnToken,
  type Location,
  type AssignmentKind,
  type DestructuringKind,
} from '../common';

export type ParserOptions = {
  shouldAddLoc?: boolean;
  shouldAddRanges?: boolean;
  /**
   * Used together with source maps. File containing the code being parsed
   */
  sourceFile?: string;
  /**
   * Holds either a function or array used on every comment
   */
  onComment?: OnComment;
  /**
   * Function invoked with the character offset when automatic semicolon insertion occurs
   */
  onInsertedSemicolon?: OnInsertedSemicolon;
  /**
   * Holds either a function or array used on every token
   */
  onToken?: OnToken;
};

export class Parser {
  private lastOnToken: [string, number, number, ESTree.SourceLocation] | null = null;

  token = Token.EOF;
  /**
   * The mutable parser flags, in case any flags need passed by reference.
   */
  flags = Flags.None;
  /**
   * The current index
   */
  index = 0;
  /**
   * Beginning of current line
   */
  line = 1;

  /**
   * Beginning of current column
   */
  column = 0;

  /**
   * Start position of whitespace/comment before current token
   */
  startIndex = 0;

  /**
   * The end of the source code
   */
  end = 0;

  /**
   * Start position of text of current token
   */
  tokenIndex = 0;

  /**
   * Start position of the column before newline
   */
  startColumn = 0;

  /**
   * Position in the input code of the first character after the last newline
   */
  tokenColumn = 0;

  /**
   * The number of newlines
   */
  tokenLine = 1;

  /**
   * Start position of text of current token
   */
  startLine = 1;

  /**
   * Holds the scanned token value
   */
  tokenValue: any = '';

  /**
   * Holds the raw text that have been scanned by the lexer
   */
  tokenRaw = '';

  /**
   * Holds the regExp info text that have been collected by the lexer
   */
  tokenRegExp: void | {
    pattern: string;
    flags: string;
  } = void 0;

  /**
   * The code point at the current index
   */
  currentChar = 0;

  /**
   *  https://tc39.es/ecma262/#sec-module-semantics-static-semantics-exportednames
   */
  exportedNames: Record<string, number> = {};

  /**
   * https://tc39.es/ecma262/#sec-exports-static-semantics-exportedbindings
   */

  exportedBindings: Record<string, number> = {};

  /**
   * Assignable state
   */
  assignable: AssignmentKind | DestructuringKind = 1;

  /**
   * Destructuring state
   */
  destructible: AssignmentKind | DestructuringKind = 0;

  /**
   * Holds leading decorators before "export" or "class" keywords
   */
  leadingDecorators: ESTree.Decorator[] = [];

  constructor(
    /**
     * The source code to be parsed
     */
    public readonly source: string,
    public readonly options: ParserOptions = {},
  ) {
    this.end = source.length;
    this.currentChar = source.charCodeAt(0);
  }

  /**
   * Get the current token in the stream to consume
   * This function exists as workaround for TS issue
   * https://github.com/microsoft/TypeScript/issues/9998
   */
  getToken() {
    return this.token;
  }

  /**
   * Set the current token in the stream to consume
   * This function exists as workaround for TS issue
   * https://github.com/microsoft/TypeScript/issues/9998
   */
  setToken(value: Token, replaceLast = false) {
    this.token = value;

    const { onToken } = this.options;

    if (onToken) {
      if (value !== Token.EOF) {
        const loc = {
          start: {
            line: this.tokenLine,
            column: this.tokenColumn,
          },
          end: {
            line: this.line,
            column: this.column,
          },
        };

        if (!replaceLast && this.lastOnToken) {
          onToken(...this.lastOnToken);
        }
        this.lastOnToken = [convertTokenType(value), this.tokenIndex, this.index, loc];
      } else {
        if (this.lastOnToken) {
          onToken(...this.lastOnToken);
          this.lastOnToken = null;
        }
      }
    }
    return value;
  }

  get tokenStart(): Location {
    return {
      index: this.tokenIndex,
      line: this.tokenLine,
      column: this.tokenColumn,
    };
  }

  finishNode<T extends ESTree.Node>(node: T, start: Location, end: Location | void): T {
    if (this.options.shouldAddRanges) {
      node.start = start.index;
      const endIndex = end ? end.index : this.startIndex;
      node.end = endIndex;
      node.range = [start.index, endIndex];
    }

    if (this.options.shouldAddLoc) {
      node.loc = {
        start: {
          line: start.line,
          column: start.column,
        },
        end: end ? { line: end.line, column: end.column } : { line: this.startLine, column: this.startColumn },
      };

      if (this.options.sourceFile) {
        node.loc.source = this.options.sourceFile;
      }
    }

    return node;
  }
}

export function pushComment(comments: ESTree.Comment[], options: ParserOptions): OnComment {
  return function (type: ESTree.CommentType, value: string, start: number, end: number, loc: ESTree.SourceLocation) {
    const comment: ESTree.Comment = {
      type,
      value,
    };

    if (options.shouldAddRanges) {
      comment.start = start;
      comment.end = end;
      comment.range = [start, end];
    }
    if (options.shouldAddLoc) {
      comment.loc = loc;
    }
    comments.push(comment);
  };
}

export function pushToken(tokens: any[], options: ParserOptions): OnToken {
  return function (type: string, start: number, end: number, loc: ESTree.SourceLocation) {
    const token: any = {
      token: type,
    };

    if (options.shouldAddRanges) {
      token.start = start;
      token.end = end;
      token.range = [start, end];
    }
    if (options.shouldAddLoc) {
      token.loc = loc;
    }
    tokens.push(token);
  };
}
