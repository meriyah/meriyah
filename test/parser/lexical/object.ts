import * as t from 'node:assert/strict';
import { describe, it } from 'vitest';
import { parseSource } from '../../../src/parser';
import { fail } from '../../test-utils';

describe('Lexical - Object', () => {
  fail('Lexical - Object (fail)', [
    { code: '!{ f(){ let a; var a; } };', options: { lexical: true } },
    { code: '!{ *g(){ let a; var a; } };', options: { lexical: true } },
    { code: '!{ get f(){ let a; var a; } };', options: { lexical: true } },
    { code: '!{ set f(b){ let a; var a; } };', options: { lexical: true } },
    { code: '!{ f(a, a){} };', options: { lexical: true } },
    { code: '!{ f([a, a]){} };', options: { lexical: true } },
    { code: '!{ f({a, a}){} };', options: { lexical: true } },
    { code: '!{ *g(a, a){} };', options: { lexical: true } },
    { code: '!{ *g([a, a]){} };', options: { lexical: true } },
    { code: '!{ *g({a, a}){} };', options: { lexical: true } },
    { code: '!{ f(a) { let a; } };', options: { lexical: true } },
    { code: '!{ f([a]){ let a; } };', options: { lexical: true } },
    { code: '!{ f({a}){ let a; } };', options: { lexical: true } },
    { code: '!{ set f({a, a}){} };', options: { lexical: true } },
    { code: '!{ set f([a, a]){} };', options: { lexical: true } },
    { code: '!{ set f(a) { let a; } };', options: { lexical: true } },
    { code: '!{ f({a, a}){} };', options: { webcompat: true, lexical: true } },
    { code: '!{ *g(a, a){} };', options: { webcompat: true, lexical: true } },
    { code: '!{ *g([a, a]){} };', options: { webcompat: true, lexical: true } },
    { code: '!{ *g({a, a}){} };', options: { webcompat: true, lexical: true } },
    { code: '!{ f(a) { let a; } };', options: { webcompat: true, lexical: true } },
    { code: '!{ f([a]){ let a; } };', options: { webcompat: true, lexical: true } },
    { code: '!{ f({a}){ let a; } };', options: { webcompat: true, lexical: true } },
    { code: '!{ set f({a, a}){} };', options: { webcompat: true, lexical: true } },
    { code: '!{ set f([a, a]){} };', options: { webcompat: true, lexical: true } },
    { code: '!{ set f(a) { let a; } };', options: { webcompat: true, lexical: true } },
    { code: '!{ set f([a]){ let a; } };', options: { lexical: true } },
    { code: '!{ set f({a}){ let a; } };', options: { lexical: true } },
    { code: '!{f(x) { let x }}', options: { lexical: true } },
    { code: '!{f(x) { const x = y }}', options: { lexical: true } },
    { code: '!{f(a, a) {}}', options: { lexical: true } },
    { code: '!{f(a, b, a) {}}', options: { lexical: true } },
    { code: '!{f(b, a, a) {}}', options: { lexical: true, sourceType: 'module' } },
    { code: '!{f(a, a, b) {}}', options: { lexical: true } },
    { code: '!{f(b, a, b, a) {}}', options: { lexical: true } },
    { code: '!{f(b, a, b, a, [fine]) {}}', options: { lexical: true } },
    { code: '!{f(b, a, b, a = x) {}}', options: { lexical: true } },
    { code: '!{f(b, a, b, ...a) {}}', options: { lexical: true } },
    { code: '!{f([a, a]) {}}', options: { lexical: true } },
    { code: '!{f([a, b, a]) {}}', options: { lexical: true } },
    { code: '!{f([b, a, a]) {}}', options: { lexical: true } },
    { code: '!{f([a, a, b]) {}}', options: { lexical: true } },
    { code: '!{f([b, a, b, a]) {}}', options: { lexical: true } },
    { code: '!{f([b, a], b) {}}', options: { lexical: true } },
    { code: '!{f([b, a], {b}) {}}', options: { lexical: true } },
    { code: '!{f([b, a], b=x) {}}', options: { lexical: true } },
    { code: '!{f([b, a], ...b) {}}', options: { lexical: true } },
    { code: '!{f(){ let x; var x; }}', options: { lexical: true } },
    { code: '!{f(){ var x; let x; }}', options: { lexical: true } },
    { code: '!{f([b, a], {b}) {}}', options: { webcompat: true, lexical: true } },
    { code: '!{f([b, a], b=x) {}}', options: { webcompat: true, lexical: true } },
    { code: '!{f([b, a], ...b) {}}', options: { impliedStrict: true, lexical: true } },
    { code: '!{f(){ let x; var x; }}', options: { sourceType: 'module', lexical: true } },
    { code: '!{f(){ var x; let x; }}', options: { sourceType: 'module', lexical: true } },
    { code: '!{f(){ const x = y; var x; }}', options: { lexical: true } },
    { code: '!{f(){ var x; const x = y; }}', options: { lexical: true } },
    { code: '!{f(){ let x; function x(){} }}', options: { lexical: true } },
    { code: '!{f(){ function x(){} let x; }}', options: { lexical: true } },
    { code: '!{f(){ const x = y; function x(){} }}', options: { lexical: true } },
    { code: '!{f(){ function x(){} const x = y; }}', options: { lexical: true } },
    { code: '!{ *foo(x) { let x = 3; } }', options: { lexical: true } },
    { code: '!{ *foo(x) { const x = 3; } }', options: { lexical: true } },
    { code: '!{f(){ const x = y; function x(){} }}', options: { sourceType: 'module', lexical: true } },
    { code: '!{f(){ function x(){} const x = y; }}', options: { sourceType: 'module', lexical: true } },
    { code: '!{ *foo(x) { let x = 3; } }', options: { sourceType: 'module', lexical: true } },
    { code: '!{ *foo(x) { const x = 3; } }', options: { sourceType: 'module', lexical: true } },
    { code: '!{f({b}, ...b) {}}', options: { lexical: true } },
    { code: '!{f({a: b}, ...b) {}}', options: { lexical: true } },
    { code: '!{f({"a": b}, ...b) {}}', options: { lexical: true } },
    { code: '!{f({2: b}, ...b) {}}', options: { lexical: true } },
    { code: '!{f({a = b}, ...a) {}}', options: { lexical: true } },
    { code: '!{f({[a]: b}, ...b) {}}', options: { lexical: true } },
    { code: '!{f({[x]: b}, b = ((b) => {}) ) {}}', options: { lexical: true } },
    { code: '!{f({[a]: b}, ...b) {}}', options: { lexical: true } },
    { code: '!{f({[a]: b}, ...b) {}}', options: { lexical: true } },
  ]);

  for (const arg of [
    '!{f(f) { }}',
    '!{f(x) { function x() {} }}',
    '!{f(){ function x(){} var x = y; }}',
    '!{f(x) { var x; }}',
  ]) {
    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`${arg}`, { lexical: true });
      });
    });

    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`${arg}`, { next: true, lexical: true });
      });
    });

    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`${arg}`);
      });
    });
  }

  for (const arg of [
    '!{f(f) { }}',
    '!{f(x) { function x() {} }}',
    '!{f(x) { var x; }}',
    '!{f(){ function x(){} var x = y; }}',
    '!{f({[a]: b}, x = ((b) => {}) ) {}}',
  ]) {
    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`${arg}`, { webcompat: true, lexical: true });
      });
    });

    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`${arg}`, { next: true, webcompat: true, lexical: true });
      });
    });
  }
});
