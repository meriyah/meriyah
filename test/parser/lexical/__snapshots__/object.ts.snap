// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Lexical - Object > Lexical - Object (fail) > !{ *foo(x) { const x = 3; } } 1`] = `
"SyntaxError [1:21-1:22]: Duplicate binding 'x'
> 1 | !{ *foo(x) { const x = 3; } }
    |                      ^ Duplicate binding 'x'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{ *foo(x) { const x = 3; } } 2`] = `
"SyntaxError [1:21-1:22]: Duplicate binding 'x'
> 1 | !{ *foo(x) { const x = 3; } }
    |                      ^ Duplicate binding 'x'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{ *foo(x) { let x = 3; } } 1`] = `
"SyntaxError [1:19-1:20]: Duplicate binding 'x'
> 1 | !{ *foo(x) { let x = 3; } }
    |                    ^ Duplicate binding 'x'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{ *foo(x) { let x = 3; } } 2`] = `
"SyntaxError [1:19-1:20]: Duplicate binding 'x'
> 1 | !{ *foo(x) { let x = 3; } }
    |                    ^ Duplicate binding 'x'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{ *g(){ let a; var a; } }; 1`] = `
"SyntaxError [1:21-1:22]: Duplicate binding 'a'
> 1 | !{ *g(){ let a; var a; } };
    |                      ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{ *g([a, a]){} }; 1`] = `
"SyntaxError [1:11-1:12]: Duplicate binding 'a'
> 1 | !{ *g([a, a]){} };
    |            ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{ *g([a, a]){} }; 2`] = `
"SyntaxError [1:11-1:12]: Duplicate binding 'a'
> 1 | !{ *g([a, a]){} };
    |            ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{ *g({a, a}){} }; 1`] = `
"SyntaxError [1:11-1:12]: Duplicate binding 'a'
> 1 | !{ *g({a, a}){} };
    |            ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{ *g({a, a}){} }; 2`] = `
"SyntaxError [1:11-1:12]: Duplicate binding 'a'
> 1 | !{ *g({a, a}){} };
    |            ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{ *g(a, a){} }; 1`] = `
"SyntaxError [1:10-1:11]: Duplicate binding 'a'
> 1 | !{ *g(a, a){} };
    |           ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{ *g(a, a){} }; 2`] = `
"SyntaxError [1:10-1:11]: Duplicate binding 'a'
> 1 | !{ *g(a, a){} };
    |           ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{ f(){ let a; var a; } }; 1`] = `
"SyntaxError [1:20-1:21]: Duplicate binding 'a'
> 1 | !{ f(){ let a; var a; } };
    |                     ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{ f([a, a]){} }; 1`] = `
"SyntaxError [1:10-1:11]: Duplicate binding 'a'
> 1 | !{ f([a, a]){} };
    |           ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{ f([a]){ let a; } }; 1`] = `
"SyntaxError [1:16-1:17]: Duplicate binding 'a'
> 1 | !{ f([a]){ let a; } };
    |                 ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{ f([a]){ let a; } }; 2`] = `
"SyntaxError [1:16-1:17]: Duplicate binding 'a'
> 1 | !{ f([a]){ let a; } };
    |                 ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{ f({a, a}){} }; 1`] = `
"SyntaxError [1:10-1:11]: Duplicate binding 'a'
> 1 | !{ f({a, a}){} };
    |           ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{ f({a, a}){} }; 2`] = `
"SyntaxError [1:10-1:11]: Duplicate binding 'a'
> 1 | !{ f({a, a}){} };
    |           ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{ f({a}){ let a; } }; 1`] = `
"SyntaxError [1:16-1:17]: Duplicate binding 'a'
> 1 | !{ f({a}){ let a; } };
    |                 ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{ f({a}){ let a; } }; 2`] = `
"SyntaxError [1:16-1:17]: Duplicate binding 'a'
> 1 | !{ f({a}){ let a; } };
    |                 ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{ f(a) { let a; } }; 1`] = `
"SyntaxError [1:15-1:16]: Duplicate binding 'a'
> 1 | !{ f(a) { let a; } };
    |                ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{ f(a) { let a; } }; 2`] = `
"SyntaxError [1:15-1:16]: Duplicate binding 'a'
> 1 | !{ f(a) { let a; } };
    |                ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{ f(a, a){} }; 1`] = `
"SyntaxError [1:9-1:10]: Duplicate binding 'a'
> 1 | !{ f(a, a){} };
    |          ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{ get f(){ let a; var a; } }; 1`] = `
"SyntaxError [1:24-1:25]: Duplicate binding 'a'
> 1 | !{ get f(){ let a; var a; } };
    |                         ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{ set f([a, a]){} }; 1`] = `
"SyntaxError [1:14-1:15]: Duplicate binding 'a'
> 1 | !{ set f([a, a]){} };
    |               ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{ set f([a, a]){} }; 2`] = `
"SyntaxError [1:14-1:15]: Duplicate binding 'a'
> 1 | !{ set f([a, a]){} };
    |               ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{ set f([a]){ let a; } }; 1`] = `
"SyntaxError [1:20-1:21]: Duplicate binding 'a'
> 1 | !{ set f([a]){ let a; } };
    |                     ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{ set f({a, a}){} }; 1`] = `
"SyntaxError [1:14-1:15]: Duplicate binding 'a'
> 1 | !{ set f({a, a}){} };
    |               ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{ set f({a, a}){} }; 2`] = `
"SyntaxError [1:14-1:15]: Duplicate binding 'a'
> 1 | !{ set f({a, a}){} };
    |               ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{ set f({a}){ let a; } }; 1`] = `
"SyntaxError [1:20-1:21]: Duplicate binding 'a'
> 1 | !{ set f({a}){ let a; } };
    |                     ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{ set f(a) { let a; } }; 1`] = `
"SyntaxError [1:19-1:20]: Duplicate binding 'a'
> 1 | !{ set f(a) { let a; } };
    |                    ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{ set f(a) { let a; } }; 2`] = `
"SyntaxError [1:19-1:20]: Duplicate binding 'a'
> 1 | !{ set f(a) { let a; } };
    |                    ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{ set f(b){ let a; var a; } }; 1`] = `
"SyntaxError [1:25-1:26]: Duplicate binding 'a'
> 1 | !{ set f(b){ let a; var a; } };
    |                          ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f(){ const x = y; function x(){} }} 1`] = `
"SyntaxError [1:29-1:30]: Duplicate binding 'x'
> 1 | !{f(){ const x = y; function x(){} }}
    |                              ^ Duplicate binding 'x'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f(){ const x = y; function x(){} }} 2`] = `
"SyntaxError [1:29-1:30]: Duplicate binding 'x'
> 1 | !{f(){ const x = y; function x(){} }}
    |                              ^ Duplicate binding 'x'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f(){ const x = y; var x; }} 1`] = `
"SyntaxError [1:25-1:26]: Duplicate binding 'x'
> 1 | !{f(){ const x = y; var x; }}
    |                          ^ Duplicate binding 'x'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f(){ function x(){} const x = y; }} 1`] = `
"SyntaxError [1:30-1:31]: Duplicate binding 'x'
> 1 | !{f(){ function x(){} const x = y; }}
    |                               ^ Duplicate binding 'x'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f(){ function x(){} const x = y; }} 2`] = `
"SyntaxError [1:30-1:31]: Duplicate binding 'x'
> 1 | !{f(){ function x(){} const x = y; }}
    |                               ^ Duplicate binding 'x'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f(){ function x(){} let x; }} 1`] = `
"SyntaxError [1:27-1:28]: Duplicate binding 'x'
> 1 | !{f(){ function x(){} let x; }}
    |                            ^ Duplicate binding 'x'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f(){ let x; function x(){} }} 1`] = `
"SyntaxError [1:23-1:24]: Duplicate binding 'x'
> 1 | !{f(){ let x; function x(){} }}
    |                        ^ Duplicate binding 'x'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f(){ let x; var x; }} 1`] = `
"SyntaxError [1:19-1:20]: Duplicate binding 'x'
> 1 | !{f(){ let x; var x; }}
    |                    ^ Duplicate binding 'x'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f(){ let x; var x; }} 2`] = `
"SyntaxError [1:19-1:20]: Duplicate binding 'x'
> 1 | !{f(){ let x; var x; }}
    |                    ^ Duplicate binding 'x'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f(){ var x; const x = y; }} 1`] = `
"SyntaxError [1:22-1:23]: Duplicate binding 'x'
> 1 | !{f(){ var x; const x = y; }}
    |                       ^ Duplicate binding 'x'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f(){ var x; let x; }} 1`] = `
"SyntaxError [1:19-1:20]: Duplicate binding 'x'
> 1 | !{f(){ var x; let x; }}
    |                    ^ Duplicate binding 'x'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f(){ var x; let x; }} 2`] = `
"SyntaxError [1:19-1:20]: Duplicate binding 'x'
> 1 | !{f(){ var x; let x; }}
    |                    ^ Duplicate binding 'x'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f([a, a, b]) {}} 1`] = `
"SyntaxError [1:9-1:10]: Duplicate binding 'a'
> 1 | !{f([a, a, b]) {}}
    |          ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f([a, a]) {}} 1`] = `
"SyntaxError [1:9-1:10]: Duplicate binding 'a'
> 1 | !{f([a, a]) {}}
    |          ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f([a, b, a]) {}} 1`] = `
"SyntaxError [1:12-1:13]: Duplicate binding 'a'
> 1 | !{f([a, b, a]) {}}
    |             ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f([b, a, a]) {}} 1`] = `
"SyntaxError [1:12-1:13]: Duplicate binding 'a'
> 1 | !{f([b, a, a]) {}}
    |             ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f([b, a, b, a]) {}} 1`] = `
"SyntaxError [1:15-1:16]: Duplicate binding 'a'
> 1 | !{f([b, a, b, a]) {}}
    |                ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f([b, a], ...b) {}} 1`] = `
"SyntaxError [1:16-1:17]: Duplicate binding 'b'
> 1 | !{f([b, a], ...b) {}}
    |                 ^ Duplicate binding 'b'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f([b, a], ...b) {}} 2`] = `
"SyntaxError [1:16-1:17]: Duplicate binding 'b'
> 1 | !{f([b, a], ...b) {}}
    |                 ^ Duplicate binding 'b'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f([b, a], {b}) {}} 1`] = `
"SyntaxError [1:14-1:15]: Duplicate binding 'b'
> 1 | !{f([b, a], {b}) {}}
    |               ^ Duplicate binding 'b'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f([b, a], {b}) {}} 2`] = `
"SyntaxError [1:14-1:15]: Duplicate binding 'b'
> 1 | !{f([b, a], {b}) {}}
    |               ^ Duplicate binding 'b'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f([b, a], b) {}} 1`] = `
"SyntaxError [1:13-1:14]: Duplicate binding 'b'
> 1 | !{f([b, a], b) {}}
    |              ^ Duplicate binding 'b'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f([b, a], b=x) {}} 1`] = `
"SyntaxError [1:13-1:14]: Duplicate binding 'b'
> 1 | !{f([b, a], b=x) {}}
    |              ^ Duplicate binding 'b'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f([b, a], b=x) {}} 2`] = `
"SyntaxError [1:13-1:14]: Duplicate binding 'b'
> 1 | !{f([b, a], b=x) {}}
    |              ^ Duplicate binding 'b'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f({"a": b}, ...b) {}} 1`] = `
"SyntaxError [1:18-1:19]: Duplicate binding 'b'
> 1 | !{f({"a": b}, ...b) {}}
    |                   ^ Duplicate binding 'b'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f({[a]: b}, ...b) {}} 1`] = `
"SyntaxError [1:18-1:19]: Duplicate binding 'b'
> 1 | !{f({[a]: b}, ...b) {}}
    |                   ^ Duplicate binding 'b'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f({[a]: b}, ...b) {}} 2`] = `
"SyntaxError [1:18-1:19]: Duplicate binding 'b'
> 1 | !{f({[a]: b}, ...b) {}}
    |                   ^ Duplicate binding 'b'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f({[a]: b}, ...b) {}} 3`] = `
"SyntaxError [1:18-1:19]: Duplicate binding 'b'
> 1 | !{f({[a]: b}, ...b) {}}
    |                   ^ Duplicate binding 'b'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f({[x]: b}, b = ((b) => {}) ) {}} 1`] = `
"SyntaxError [1:16-1:17]: Duplicate binding 'b'
> 1 | !{f({[x]: b}, b = ((b) => {}) ) {}}
    |                 ^ Duplicate binding 'b'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f({2: b}, ...b) {}} 1`] = `
"SyntaxError [1:16-1:17]: Duplicate binding 'b'
> 1 | !{f({2: b}, ...b) {}}
    |                 ^ Duplicate binding 'b'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f({a = b}, ...a) {}} 1`] = `
"SyntaxError [1:17-1:18]: Duplicate binding 'a'
> 1 | !{f({a = b}, ...a) {}}
    |                  ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f({a: b}, ...b) {}} 1`] = `
"SyntaxError [1:16-1:17]: Duplicate binding 'b'
> 1 | !{f({a: b}, ...b) {}}
    |                 ^ Duplicate binding 'b'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f({b}, ...b) {}} 1`] = `
"SyntaxError [1:13-1:14]: Duplicate binding 'b'
> 1 | !{f({b}, ...b) {}}
    |              ^ Duplicate binding 'b'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f(a, a) {}} 1`] = `
"SyntaxError [1:8-1:9]: Duplicate binding 'a'
> 1 | !{f(a, a) {}}
    |         ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f(a, a, b) {}} 1`] = `
"SyntaxError [1:8-1:9]: Duplicate binding 'a'
> 1 | !{f(a, a, b) {}}
    |         ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f(a, b, a) {}} 1`] = `
"SyntaxError [1:11-1:12]: Duplicate binding 'a'
> 1 | !{f(a, b, a) {}}
    |            ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f(b, a, a) {}} 1`] = `
"SyntaxError [1:11-1:12]: Duplicate binding 'a'
> 1 | !{f(b, a, a) {}}
    |            ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f(b, a, b, ...a) {}} 1`] = `
"SyntaxError [1:17-1:18]: Duplicate binding 'a'
> 1 | !{f(b, a, b, ...a) {}}
    |                  ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f(b, a, b, a = x) {}} 1`] = `
"SyntaxError [1:15-1:16]: Duplicate binding 'a'
> 1 | !{f(b, a, b, a = x) {}}
    |                ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f(b, a, b, a) {}} 1`] = `
"SyntaxError [1:14-1:15]: Duplicate binding 'a'
> 1 | !{f(b, a, b, a) {}}
    |               ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f(b, a, b, a, [fine]) {}} 1`] = `
"SyntaxError [1:14-1:15]: Duplicate binding 'a'
> 1 | !{f(b, a, b, a, [fine]) {}}
    |               ^ Duplicate binding 'a'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f(x) { const x = y }} 1`] = `
"SyntaxError [1:17-1:18]: Duplicate binding 'x'
> 1 | !{f(x) { const x = y }}
    |                  ^ Duplicate binding 'x'"
`;

exports[`Lexical - Object > Lexical - Object (fail) > !{f(x) { let x }} 1`] = `
"SyntaxError [1:15-1:16]: Duplicate binding 'x'
> 1 | !{f(x) { let x }}
    |                ^ Duplicate binding 'x'"
`;
